sr_finger_mount/src/sr_finger_mount.py

```diff
@@ -0,0 +1,178 @@
+
+# Copyright 2021 Shadow Robot Company Ltd.
+#
+# This program is free software: you can redistribute it and/or modify it
+# under the terms of the GNU General Public License as published by the Free
+# Software Foundation version 2 of the License.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program. If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import absolute_import
+from __future__ import division
+import argparse
+import sys
+import math
+
+import rospy
+import numpy as np
+import sounddevice as sd
+import matplotlib.pyplot as plt
+from sr_robot_msgs.msg import ShadowPST
+import threading
+import time
+
+
+class DeviceHandler(threading.Thread):
+    def __init__(self, device, fingers, mount):
+        super(DeviceHandler, self).__init__()
+        self.device_name = device
+        self.fingers = fingers
+        self.mount = mount
+        self.start_idx = [0] * len(self.fingers)
+        self.m_phase = [0] * len(self.fingers)
+        self.oldsignal = [0] * len(self.fingers)
+        self.freq = [1] * len(self.fingers)
+        self.amp = [1] * len(self.fingers)
+
+    def run(self):
+        '''
+        # FOR-LOOP ONLY FOR PLOTTING PURPOSES
+        for f in self.fingers:
+            self.mount._collection[f] = list()
+        '''
+        self.start_piezo(self.fingers)
+
+    def callback(self, outdata, frames, time, status):
+        if status:
+            print(status, file=sys.stderr)
+
+        for i, finger in enumerate(self.fingers):
+
+            t = (self.start_idx[i] + np.arange(frames)) / self.samplerate
+            t = t.reshape(-1, 1)
+            self.start_idx[i] += frames
+
+            for frame in range(frames):
+                phaseInc = 2*np.pi*self.freq[i]/self.samplerate
+                outdata[frame, i] = self.amp[i] * np.sin(self.m_phase[i])
+                self.m_phase[i] += phaseInc
+
+                if np.sign(outdata[frame, i]) != np.sign(self.oldsignal[i]):
+                    self.freq[i] = self.mount._fm[finger]
+                    self.amp[i] = self.mount._am[finger]
+                self.oldsignal[i] = outdata[frame, i]
+            '''
+            # ONLY FOR PLOTTING PURPOSES
+            self.mount._collection[finger].extend(outdata[:, i])
+            '''
+
+    def start_piezo(self, fingers):
+        self.samplerate = sd.query_devices(self.device_name, 'output')['default_samplerate']
+        with sd.OutputStream(device=self.device_name, channels=len(fingers), callback=self.callback,
+                             samplerate=self.samplerate):
+            while not rospy.is_shutdown():
+                continue
+            '''
+            # ONLY FOR PLOTTING PURPOSES
```

Comment:
delete? or is this ever useful actually? otherwise we should parametrize

Reply:
Only usefull during testing. Removed